In this homework, I want to find the likehood of appearance for the five most popular topics in New York Times obituaries corpus in the 1947 Henry Ford's text. Personally, I like Henry Ford very much, so I want to know what kind of topics he likes to write about the most.

In [22]:
#Pre-processing of data

path_to_mallet = '/data/mallet/bin/mallet'

!pip install little_mallet_wrapper #install packages
!pip install seaborn

import little_mallet_wrapper  #install packages
import seaborn
import glob
from pathlib import Path

directory = "./docs/NYT-Obituaries/"
files = glob.glob(f"{directory}/*.txt")
#Get Training Data

In [3]:
#Stopword removal
training_data = []  #pre-process the data and add to our master list training_data
for file in files:
    text = open(file, encoding='utf-8').read()
    processed_text = little_mallet_wrapper.process_string(text, numbers='remove')
    training_data.append(processed_text)

#Make a master list of the original text of the obituaries for future reference.
original_texts = []
for file in files:
    text = open(file, encoding='utf-8').read()
    original_texts.append(text)

#process titles
obit_titles = [Path(file).stem for file in files]
obit_titles

In [27]:
little_mallet_wrapper.print_dataset_stats(training_data) #get trainning data stats

num_topics = 10
training_data = training_data

Number of Documents: 378
Mean Number of Words: 1316.5
Vocabulary Size: 35945


In [28]:


#Change to your desired output directory
output_directory_path = 'topic-model-output/NYT-Obits'

#No need to change anything below here
Path(f"{output_directory_path}").mkdir(parents=True, exist_ok=True)

path_to_training_data           = f"{output_directory_path}/training.txt"
path_to_formatted_training_data = f"{output_directory_path}/mallet.training"
path_to_model                   = f"{output_directory_path}/mallet.model.{str(num_topics)}"
path_to_topic_keys              = f"{output_directory_path}/mallet.topic_keys.{str(num_topics)}"
path_to_topic_distributions     = f"{output_directory_path}/mallet.topic_distributions.{str(num_topics)}"

little_mallet_wrapper.import_data(path_to_mallet,
                path_to_training_data,
                path_to_formatted_training_data,
                training_data)
little_mallet_wrapper.train_topic_model(path_to_mallet,
                      path_to_formatted_training_data,
                      path_to_model,
                      path_to_topic_keys,
                      path_to_topic_distributions,
                      num_topics)

Importing data...
Complete
Training topic model...

Mallet LDA: 10 topics, 4 topic bits, 1111 topic mask
Data loaded.
max tokens: 19800
total tokens: 497654
<10> LL/token: -10.09804
<20> LL/token: -9.72505
<30> LL/token: -9.59621
<40> LL/token: -9.53044

0	0.5	great one time years could world many life made business never country company man day age however found gave england 
1	0.5	miss said years york first film won made one three year show career television new played stage theater movie became 
2	0.5	general war united states army military american foreign hitler nations truman soviet union took secretary president germany german world first 
3	0.5	government party political peace war france british would minister congress french later king power israel said people made leader britain 
4	0.5	new times one york wrote life grant story made book work home time last writing city came read old death 
5	0.5	first mao band man jazz came director actor group james stalin could play several

In [29]:
topic_distributions = little_mallet_wrapper.load_topic_distributions(path_to_topic_distributions)
topics = little_mallet_wrapper.load_topic_keys(path_to_topic_keys)

In [30]:
obituary_to_check = "1947-Henry-Ford"

obit_number = obit_titles.index(obituary_to_check)

print(f"Topic Distributions for {obit_titles[obit_number]}\n")
for topic_number, (topic, topic_distribution) in enumerate(zip(topics, topic_distributions[obit_number])):
    print(f"✨Topic {topic_number} {topic} ✨\nProbability: {round(topic_distribution, 3)}\n")

Topic Distributions for 1947-Henry-Ford

✨Topic 0 ['one', 'made', 'time', 'years', 'new', 'great', 'would', 'first', 'came', 'could', 'world', 'day', 'man', 'went', 'two', 'took', 'old', 'long', 'many', 'home'] ✨
Probability: 0.677

✨Topic 1 ['first', 'film', 'won', 'movie', 'year', 'broadway', 'films', 'miss', 'york', 'new', 'hollywood', 'theater', 'two', 'made', 'stage', 'movies', 'actor', 'later', 'years', 'times'] ✨
Probability: 0.001

✨Topic 2 ['war', 'soviet', 'united', 'general', 'party', 'military', 'army', 'states', 'american', 'mao', 'communist', 'hitler', 'china', 'union', 'world', 'political', 'power', 'german', 'french', 'germany'] ✨
Probability: 0.083

✨Topic 3 ['king', 'israel', 'british', 'minister', 'peace', 'government', 'war', 'gandhi', 'queen', 'prime', 'prince', 'india', 'first', 'sir', 'england', 'lord', 'britain', 'arab', 'churchill', 'year'] ✨
Probability: 0.0

✨Topic 4 ['times', 'grant', 'gen', 'general', 'book', 'new', 'york', 'published', 'first', 'story', 'w

Since the first six key words don't tell a lot difference, I chose the whole 20 key words, and find that Henry Ford writes about the story in general life between men and women with the high probability of 0.677.

In the next step, we could analyze the result with his years and contexts. Or we can do topic modeling for other writers, compare them together, and find trends in topics for 20th century American literature. 